File: basics/quick_sort.cpp
void quick_sort(vector<int> nums) {
  qsort(nums, 0, nums.length() - 1);
}

void qsort(vector<int>& nums, int left, int right) {
  if (left >= right) return;

  int M = left + (right - left)/2;
  swap(nums[left], nums[M]);
  int pivot = nums[left], L = left + 1, R = right;

  while (L <= R) {
    while (L <= R && nums[L] < pivot) ++ L;
    while (L <= R && nums[R] > pivot) -- R;
    if (L < R) swap(nums[L ++], nums[R --]);
    else ++ L;
  }

  swap(nums[left], nums[R]);
  qsort(nums, left, R - 1);
  qsort(nums, R + 1, right);
}


void quick_select(vector<int>& nums, int left, int right, int k) {
  if (left >= right) return;

  int M = left + (right - left)/2;
  swap(nums[left], nums[M]);
  int pivot = nums[left], L = left + 1, R = right;

  while (L <= R) {
    while (L <= R && nums[L] < pivot) ++ L;
    while (L <= R && nums[R] > pivot) -- R;
    if (L < R) swap(nums[L ++], nums[R --]);
    else ++ L;
  }

  swap(nums[left], nums[R]);
  if (k < R) quick_select(nums, left, R - 1, k);
  if (k > R) quick_select(nums, R + 1, right, k);
}

File: basics/quick_sort.py

    def quickSort(arr, left, right, k):
        if left >= right:
            return
        
        mid = left + (right - left) // 2
        arr[left], arr[mid] = arr[mid], arr[left]
        pivot = arr[left]
        L, R = left + 1, right

        while L <= R:
            while L <= R and arr[L] <= pivot: 
                L += 1
            while L <= R and arr[R] >= pivot: 
                R -= 1
            if L < R:
                arr[L], arr[R] = arr[R], arr[L]
                L += 1
                R -= 1
            else:
                L += 1
        
        arr[left], arr[R] = arr[R], arr[left]
        quickSort(arr, left, R-1, k)
        quickSort(arr, R+1, right, k)


    def quickSelect(arr, left, right, k):
        if left >= right:
            return
        
        mid = left + (right - left) // 2
        arr[left], arr[mid] = arr[mid], arr[left]
        pivot = arr[left]
        L, R = left + 1, right

        while L <= R:
            while L <= R and arr[L] <= pivot: 
                L += 1
            while L <= R and arr[R] >= pivot: 
                R -= 1
            if L < R:
                arr[L], arr[R] = arr[R], arr[L]
                L += 1
                R -= 1
            else:
                L += 1
        
        arr[left], arr[R] = arr[R], arr[left]
        if k < R:
            self.quickSelect(arr, left, R-1, k)
        elif k > R:
            self.quickSelect(arr, R+1, right, k)
